File: third_site/php/spirit.php
<?php 
define("SPIRIT_HOST", "https://www.printspirit.cn");
#define("SPIRIT_HOST", "http://192.168.1.100:8059");
define("UID",  "third_test");
define("PASS", "third_test");

function getAccessToken($uid, $pass) {

	$apcuAvailabe = function_exists('apcu_enabled') && apcu_enabled();
	
	if($apcuAvailabe){
		$access_token = apcu_fetch('access_token');
		$expirt_time = apcu_fetch('expirt_time');
		if ( $access_token && $expirt_time > time() ) return $access_token;
	}

	$rc=json_decode(file_get_contents(SPIRIT_HOST . "/api/get-access-token?userid=$uid&passwd=$pass"));

	if ($rc!=NULL && $rc->rc=='OK') {
		if($apcuAvailabe){
			apcu_store('access_token', $rc->token);
			apcu_store('expirt_time',  time() + $rc->expirt);
		}
		return $rc->token;
	}	
	die("无法获取TOKEN:".$rc->errmsg);
}

function getList($subclass="") {
	$token = getAccessToken(UID, PASS);
	$rc=json_decode(file_get_contents(SPIRIT_HOST . "/api/get-label-list?token=${token}&subclass=${subclass}"), true);
	if ($rc!=NULL && $rc['rc']=='OK') 	return $rc['data'];
	return [];
}

function getContent($tpid) {
	$token = getAccessToken(UID, PASS);
	$rc=json_decode(file_get_contents(SPIRIT_HOST . "/api/get-label-content?token=${token}&tpid=${tpid}"), true);
	if ($rc!=NULL && $rc['rc']=='OK') return $rc['data'];
	return "";
}

function get_edit_url($subclass, $tpid="") {
	$token = getAccessToken(UID, PASS);
	return SPIRIT_HOST . "/third-edit?subclass=${subclass}&tpid=${tpid}&token=${token}";	
}

File: third_site/go/app/spirit.go
package app

import (
	"encoding/json"
	"fmt"
	"net/http"
	"time"
)

var (
	SPIRIT_HOST = "https://www.printspirit.cn"
)

type ThirdApp struct {
	Host, Uid, Pass    string 
	Token        string
	Expired_time int64
}

func NewThirdApp(host, uid, pass string) *ThirdApp {
	return &ThirdApp{
		Host: host,
		Uid : uid, 
		Pass: pass,
	}
}

func (app *ThirdApp) getAccessToken() (string, error) {

	if app.Expired_time > time.Now().Unix() {
		return app.Token, nil
	}
	url := fmt.Sprintf("%s/api/get-access-token?userid=%v&passwd=%s", app.Host, app.Uid, app.Pass)

	resp, err := http.Get(url)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	rc := make(map[string]interface{})
	err = json.NewDecoder(resp.Body).Decode(&rc)
	if err != nil {
		return "", err
	}
	if rc["rc"].(string) != "OK" {
		return "", fmt.Errorf(rc["errmsg"].(string))
	}
	app.Expired_time = time.Now().Unix() + int64(rc["expirt"].(float64))
	app.Token = rc["token"].(string)
	return app.Token, nil
}

type TpInfo struct {
	Name     string `json:"name"`
	Subclass string `json:"subclass"`
	Id       string `json:"id"`
}

func (app *ThirdApp) GetList(subclass string) (*[]TpInfo, error) {
	token, err := app.getAccessToken()
	if err != nil {
		return nil, err
	}
	url := fmt.Sprintf("%s/api/get-label-list?token=%s&subclass=%s", app.Host, token, subclass)

	resp, err := http.Get(url)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	rc := struct {
		Rc     string   `json:"rc"`
		Errmsg string   `json:"errmsg"`
		Data   []TpInfo `json:"data"`
	}{}
	err = json.NewDecoder(resp.Body).Decode(&rc)
	if err != nil {
		return nil, err
	}
	if rc.Rc != "OK" {
		return nil, fmt.Errorf(rc.Errmsg)
	}
	return &rc.Data, nil
}

func (app *ThirdApp) GetEditUrl(subclass, tpid string) (string, error) {
	token, err := app.getAccessToken()
	if err != nil {
		return "", err
	}
	return fmt.Sprintf("%s/third-edit?subclass=%s&tpid=%s&token=%s", app.Host, subclass, tpid, token), nil
}
